### PR TITLE
125 update roadmap

### LINKED ISSUE #125: update roadmap
Describe the bug
Current roadmap section in readme.md, while important, is incomplete.

Additional context
Add more plans for the future, based on discusord discussion in private group with @fkiraly and @benHeid

### PR DESCRIPTION
**Pull request by astrogilda**
# Pull Request Template

## Description

Updated `roadmap` in `README.md`. 
closes #125 
Expanded roadmap at #144

## Type of change

Please delete options that are not relevant.

- [ x] This change requires a documentation update

## Checklist:

- [ x] My code follows the style guidelines of this project
- [ x] I have performed a self-review of my own code
- [ x] I have made corresponding changes to the documentation
- [x ] My changes generate no new warnings

## Add All Contributors Command

Remember to acknowledge your contributions, replace `contribution_type` with your contribution (code, doc, etc.):

```plaintext
@all-contributors please add @<your-username> for <contribution_type>
```

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
It seems like this PR accidentally contains commits from multiple PR. Accidental push?

Can you kindly untangle this?

**Comment by astrogilda** [comment]
> It seems like this PR accidentally contains commits from multiple PR. Accidental push?
> 
> Can you kindly untangle this?

Just rebased against main, should be g2g!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -108,4 +108,4 @@ jobs:
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           password: ${{ secrets.PYPI_TOKEN }}
-          packages_dir: wheelhouse/
+          packages-dir: wheelhouse/
```
Comment (fkiraly) [inline .github/workflows/release.yml:111]: unintended?

File: src/tsbootstrap/base_bootstrap.py [hunk 2]
```diff
@@ -25,7 +25,6 @@
     OrderTypes,
 )
 
-
```
Comment (fkiraly) [inline src/tsbootstrap/base_bootstrap.py:28]: I think this will fail flake8 linting

**Review by fkiraly — CHANGES_REQUESTED** [review]
It still seems there are multiple changes in this PR unrelated to the roadmap, some of them appearing accidental?

I suggest moving anything that is not adding roadmap into separate PR and commenting, so at least it's clear that the changes are deliberate and not accidental.
(of course you can make any changes you like to your repo, I am just flagging the risk of accidentally breaking things without review)

Regarding the roadmap: I think it makes sense overall.
I feel it is a bit verbose though - personally, I prefer bullet point lists, with more extensive text possibly in a second section, or design documents.

I have also some ideas which are not necessary to add, but for consideration.

**Comment by fkiraly** [comment]
> Just rebased against main, should be g2g!

My recommendation is merging from main. I know there are differring schools here, I just feel the risk for unintended reverts (like the above?) is less if you do things quickly.

**Comment by fkiraly** [comment]
Some random ideas for roadmap:

### integration

* using arbitrary `sktime` forecaster in forecaster based bootstraps
* using `tsbootstrap` bootstraps to create a distribution or sampler-like object (`skpro` or `sktime` proba distribution over time series)


### boilerplate/sugar

* fractions for block lengths etc
* ability to pass `pd.DataFrame`. "core" behaviour is `np.ndarray`, but if `pd.DataFrame` is passed, produces the output to have same index and columns


### refactor

* there is some legacy entanglement of classes. These should be refactored to be "flatter".
* the block bootstraps look like they should be composites


### API extension

* update/stream
* panels and hierarchical data
* exogenous data (API supports but there is no logic except for forecasters)
* fittable or pretrained


### adjacent areas

* time series augmentation
* full proba models that can be sampled from
* time series data generation

File: .github/workflows/release.yml (hunk 1 above)
Comment (astrogilda) [reply .github/workflows/release.yml:111]: nah, `ruff`s recommendation

File: README.md [hunk 3]
```diff
@@ -316,7 +316,7 @@ n_samples = 1000
 
 y = np.random.normal(0, 1, n_samples).cumsum()
 
-x1 = np.arange(1, n_samples + 1).reshape(-1, 1)
+x1 = np.arrange(1, n_samples + 1).reshape(-1, 1)
```
[the remark this answers is not in the document]
Comment (astrogilda) [reply README.md:319]: yep! reverted

**Review by fkiraly — CHANGES_REQUESTED** [review]
still something weird going on with all files being changed.

**Comment by astrogilda** [comment]
Hmm this is after reading against main.
Maybe let's focus on the readme.md file for now given the time constraints?

File: .github/workflows/release.yml (hunk 1 above)
Comment (fkiraly) [reply .github/workflows/release.yml:111]: are both keywords valid?

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply README.md:319]: does not look like it

**Review by fkiraly — CHANGES_REQUESTED** [review]
various changes issue still there, I think it's better to split.

Re roadmap, I think it would be better to move too verbose content somewher else, and have only an extra short bullet point list in the README.

One option is docs, another option is an issue, for 2024-2025 roadmap.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks, and for resolving the merge conflicts!

My suggestion would be to split this roadmap in two: a high-level, short part in the README, and an issue ("tsbootstrap roadmap 2024-2025" or similar) which the README links to.

E.g., high-level as follows

* **performance and scaling** - handling large datasets, distributed backend integration (dask, spark, ray), profiling/optimization
* **tuning and autoML** - adaptive block length, adaptive resampling, evaluation based parameter selection
* **real-time and stream data** - stream bootstraps, data update interface
* **stage 2 `sktime` integration** - evaluation module, datasets, benchmarks, `sktime` forecasters in bootstraps
* **API and capability extension** - panel/hierarchical data, exogenous data, update/stream, model state management
* **scope extension** (to be discussed) - time series augmentation, fully probabilistic models

Further details: [link to decompressed version]

**Review by fkiraly — APPROVED** [review]
Yes, that's nice - I also liked the longer version though, it had more detail.

I would still suggest to open an issue and link.

**Comment by astrogilda** [comment]
it's linked in the updated readme.py

**Comment by astrogilda** [comment]
Just tried, I guess you can't. I'll put it here.

**Comment by fkiraly** [comment]
you could just add an ordinary link, like `[name](URL)`

**Pull request merged**